File: packages/firebase_core/ios/Classes/FirebaseCorePlugin.m
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "FirebaseCorePlugin.h"

#import <Firebase/Firebase.h>

static NSDictionary *getDictionaryFromFIROptions(FIROptions *options) {
  return @{
    @"googleAppID" : options.googleAppID ?: [NSNull null],
    @"bundleID" : options.bundleID ?: [NSNull null],
    @"GCMSenderID" : options.GCMSenderID ?: [NSNull null],
    @"APIKey" : options.APIKey ?: [NSNull null],
    @"clientID" : options.clientID ?: [NSNull null],
    @"trackingID" : options.trackingID ?: [NSNull null],
    @"projectID" : options.projectID ?: [NSNull null],
    @"androidClientID" : options.androidClientID ?: [NSNull null],
    @"databaseUrl" : options.databaseURL ?: [NSNull null],
    @"storageBucket" : options.storageBucket ?: [NSNull null],
    @"deepLinkURLScheme" : options.deepLinkURLScheme ?: [NSNull null],
  };
}

static NSDictionary *getDictionaryFromFIRApp(FIRApp *app) {
  return @{@"name" : app.name, @"options" : getDictionaryFromFIROptions(app.options)};
}

@implementation FLTFirebaseCorePlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  FlutterMethodChannel *channel =
      [FlutterMethodChannel methodChannelWithName:@"plugins.flutter.io/firebase_core"
                                  binaryMessenger:[registrar messenger]];
  FLTFirebaseCorePlugin *instance = [[FLTFirebaseCorePlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
  if ([@"FirebaseApp#configure" isEqualToString:call.method]) {
    NSString *name = call.arguments[@"name"];
    NSDictionary *optionsDictionary = call.arguments[@"options"];
    FIROptions *options =
        [[FIROptions alloc] initWithGoogleAppID:optionsDictionary[@"googleAppID"]
                                    GCMSenderID:optionsDictionary[@"GCMSenderID"]];
    if (![optionsDictionary[@"bundleID"] isEqual:[NSNull null]])
      options.bundleID = optionsDictionary[@"bundleID"];
    if (![optionsDictionary[@"APIKey"] isEqual:[NSNull null]])
      options.APIKey = optionsDictionary[@"APIKey"];
    if (![optionsDictionary[@"clientID"] isEqual:[NSNull null]])
      options.clientID = optionsDictionary[@"clientID"];
    if (![optionsDictionary[@"trackingID"] isEqual:[NSNull null]])
      options.trackingID = optionsDictionary[@"trackingID"];
    if (![optionsDictionary[@"projectID"] isEqual:[NSNull null]])
      options.projectID = optionsDictionary[@"projectID"];
    if (![optionsDictionary[@"androidClientID"] isEqual:[NSNull null]])
      options.androidClientID = optionsDictionary[@"androidClientID"];
    if (![optionsDictionary[@"databaseURL"] isEqual:[NSNull null]])
      options.databaseURL = optionsDictionary[@"databaseURL"];
    if (![optionsDictionary[@"storageBucket"] isEqual:[NSNull null]])
      options.storageBucket = optionsDictionary[@"storageBucket"];
    if (![optionsDictionary[@"deepLinkURLScheme"] isEqual:[NSNull null]])
      options.deepLinkURLScheme = optionsDictionary[@"deepLinkURLScheme"];
    [FIRApp configureWithName:name options:options];
    result(nil);
  } else if ([@"FirebaseApp#allApps" isEqualToString:call.method]) {
    NSDictionary<NSString *, FIRApp *> *allApps = [FIRApp allApps];
    NSMutableArray *appsList = [NSMutableArray array];
    for (NSString *name in allApps) {
      FIRApp *app = allApps[name];
      [appsList addObject:getDictionaryFromFIRApp(app)];
    }
    result(appsList.count > 0 ? appsList : nil);
  } else if ([@"FirebaseApp#appNamed" isEqualToString:call.method]) {
    NSString *name = call.arguments;
    FIRApp *app = [FIRApp appNamed:name];
    result(getDictionaryFromFIRApp(app));
  } else {
    result(FlutterMethodNotImplemented);
  }
}

@end

File: packages/firebase_core/android/src/main/java/io/flutter/plugins/firebase/core/FirebaseCorePlugin.java
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.firebase.core;

import android.content.Context;
import com.google.firebase.FirebaseApp;
import com.google.firebase.FirebaseOptions;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;
import java.lang.String;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class FirebaseCorePlugin implements MethodCallHandler {

  private final Context context;

  public static void registerWith(PluginRegistry.Registrar registrar) {
    final MethodChannel channel =
        new MethodChannel(registrar.messenger(), "plugins.flutter.io/firebase_core");
    channel.setMethodCallHandler(new FirebaseCorePlugin(registrar.context()));
  }

  private FirebaseCorePlugin(Context context) {
    this.context = context;
  }

  private Map asMap(FirebaseApp app) {
    Map<String, Object> appMap = new HashMap<>();
    appMap.put("name", app.getName());
    FirebaseOptions options = app.getOptions();
    Map<String, String> optionsMap = new HashMap<>();
    optionsMap.put("googleAppID", options.getApplicationId());
    optionsMap.put("GCMSenderID", options.getGcmSenderId());
    optionsMap.put("APIKey", options.getApiKey());
    optionsMap.put("databaseURL", options.getDatabaseUrl());
    optionsMap.put("storageBucket", options.getStorageBucket());
    optionsMap.put("projectID", options.getProjectId());
    appMap.put("options", optionsMap);
    return appMap;
  }

  @Override
  public void onMethodCall(MethodCall call, final Result result) {
    switch (call.method) {
      case "FirebaseApp#configure":
        {
          Map<String, Object> arguments = call.arguments();
          String name = (String) arguments.get("name");
          @SuppressWarnings("unchecked")
          Map<String, String> optionsMap = (Map<String, String>) arguments.get("options");
          FirebaseOptions options =
              new FirebaseOptions.Builder()
                  .setApiKey(optionsMap.get("APIKey"))
                  .setApplicationId(optionsMap.get("googleAppID"))
                  .setDatabaseUrl(optionsMap.get("databaseURL"))
                  .setGcmSenderId(optionsMap.get("GCMSenderID"))
                  .setProjectId(optionsMap.get("projectID"))
                  .setStorageBucket(optionsMap.get("storageBucket"))
                  .build();
          FirebaseApp.initializeApp(context, options, name);
          result.success(null);
          break;
        }
      case "FirebaseApp#allApps":
        {
          List<Map<String, Object>> apps = new ArrayList<>();
          for (FirebaseApp app : FirebaseApp.getApps(context)) {
            apps.add(asMap(app));
          }
          result.success(apps);
          break;
        }
      case "FirebaseApp#appNamed":
        {
          String name = (String) call.arguments();
          try {
            FirebaseApp app = FirebaseApp.getInstance(name);
            result.success(asMap(app));
          } catch (IllegalStateException ex) {
            // App doesn't exist, so successfully return null.
            result.success(null);
          }
          break;
        }
      default:
        {
          result.notImplemented();
          break;
        }
    }
  }
}
